indirect.h

```diff
@@ -20,37 +20,33 @@ class indirect {
   indirect() = default;
 
   template <class... Ts>
-  indirect(std::in_place_t, Ts&& ...ts) {
+  indirect(std::in_place_t, Ts&&... ts) {
     ptr_ = std::unique_ptr<T, D>(new T(std::forward<Ts>(ts)...), D{});
   }
 
   indirect(T* t, C c = C{}, D d = D{}) : c_(std::move(c)) {
     ptr_ = std::unique_ptr<T, D>(t, std::move(d));
   }
 
-  indirect(const indirect& i) {
-    if (i.ptr_) { 
+  indirect(const indirect& i) : c_(i.c_) {
+    if (i.ptr_) {
```

Comment:
It seems that copy constructor and assignment operator both have to duplicate logic to deal with the conceptual problem of "only call the copy operations if the right-hand side of the assignment is not null".  This makes sense because the call operator accepts a const reference (as the underlying copy constructor it invokes also does).  However, this complicates the logic of all call sites.  We can't move this inside the copier object call operator as these can be user-defined and that adds complexity to the logic for any implementers.  So I'd suggest we add a wrapper function:
```cpp
T* copy(T* rhs)
{
    if (rhs)
    {
        return c_(*rhs);
    }
    return nullptr;
}
```

This should simplify the 
```
indirect(const indirect& i) : c_(i.c_), ptr_(copy(i.ptr_)) {}

indirect& operator=(const indirect& i) {
    c_ = i.c_;
    ptr_ = std::unique_ptr<T, D>(copy(i.ptr_), D{});
}
```
This will remove a few inconsistencies I see with the current implementation, that if the right-hand side is null it is ignored, in which case we copy the copier object, but not the deleter object?  While it seems inconsistent, it probably isn't important because for the copier to be used, a second successful copy operation has to be invoked in which case it will be copied over again before being used anyway (so the copy/assignment of the copier object could have been moved in the if).

Using the wrapper copy function we also avoid the issue of double initialising ptr_.   This allows the copy to happen via the initialiser list, which then avoids the default constructor for ptr_ being called implicitly before the actual body of the constructor (which is causing duplicated work in the case the pointer is initialised to a valid value).  Hope that is all clear.